### DISCUSSION TITLE
[Proposal]: Bulk copy properties from one object to another

### DISCUSSION THREAD

**Discussion by NETKroks** [dotnet/csharplang#8082] [Language Ideas]
One potential Use-Case: Updating an Entity by using Entity Framework. I do not want to type this pattern out for every property:
```csharp
obj1.PropertyOne = obj2.PropertyOne;
obj1.PropertyTwo = obj2.PropertyTwo;
...
```

Another one would be: you keep track of objects in-memory and you provide an Update API. The User can pass an object with the same type to the method and you want to update all properties.

More explicit example using ASP:
```csharp
    [HttpPatch("...")]
    public async Task<IActionResult> UpdateCustomType([FromBody(EmptyBodyBehavior = EmptyBodyBehavior.Disallow)] CustomType obj2)
    {
        var obj1 = CustomTypeManager.GetById(obj2.Id);
        
        obj1.PropertyOne = obj2.PropertyOne;
        obj1.PropertyTwo = obj2.PropertyTwo ;
        ...
        
        return Ok();
    }
```
In this example I may not be able just to replace the entire object since the object could be referenced multiple times by other things. The reference would not be available anymore.

We could instead introduce `Unsafe.CopyProperties<T>(from: t1, to: t2)` which would basically bulk copy all field values from obj1 to obj2 (thus the properties are changed since they just get and set the backing fields). Since obj1 and obj2 are equal in type, we can even assume its safe to copy raw bytes over to the other object.
This would make it a lot more convenient as you do not need to type out every property. And it would be faster than the reflection approach.

With this the code above would look like:
```csharp
    [HttpPatch("...")]
    public async Task<IActionResult> UpdateCustomType([FromBody(EmptyBodyBehavior = EmptyBodyBehavior.Disallow)] CustomType obj2)
    {
        var obj1 = CustomTypeManager.GetById(obj2.Id);
        Unsafe.CopyProperties(from: obj2, to: obj1);
        return Ok();
    }
```

**Comment by KennethHoff** [comment]
Duplicates https://github.com/dotnet/csharplang/issues/162 ?

**Reply by iam3yal** [reply]
It's not, he asks for an API to be added and the semantics seem to be very different.

**Comment by iam3yal** [comment]
@NETKroks This does not require a language change and should probably be posted [here](https://github.com/dotnet/runtime).

**Reply by NETKroks** [reply]
My bad, surely will do.

**Comment by NETKroks** [comment]
The issue is discussed here: https://github.com/dotnet/runtime/issues/101676

**Comment by HaloFour** [comment]
Sounds like a job for source generators.

**Reply by HaloFour** [reply]
Looks like one already exists: https://github.com/riok/mapperly
